### DISCUSSION TITLE
Some feedback

### DISCUSSION THREAD

**Discussion by SirLynix** [jrouwe/JoltPhysics#493] [General]
Following #486, I though it would be more interesting to open a discussion to give more feedback.

I've been using [Newton Dynamics](https://github.com/MADEAPPS/newton-dynamics) for years, it's a neat physics engine which deserve to be known but it hasn't been tested on a lot of platforms and I've run into trouble to compile it on Android/iOS/emscripten.

I used [Bullet](https://github.com/bulletphysics/bullet3) next, which runs on all platforms, has some nice demo and soft bodies support, however it wasn't as straightforward as Newton to setup and is a bit confusing (for example should one use src/Bullet3Dynamics or src/BulletDynamics? Why are both present in the same branch?). Also some of its code looks like it could be improved (for example removing a body from the world performs a linear search in a big array containing all moving bodies).

And then I tried your engine, Jolt, which has great examples but a more complicated code, more game-oriented or should I say, professionally looking.

Setuping Jolt seemed hard in the first place because of all the things you have to setup yourself, why are there so many globals you have to set yourself? Why do you have to take care of giving it a threadpool and a temp allocator instead of it handling it by itself, like Bullet/Newton does?

After a few days of experimenting with Jolt it became clear this was to integrate in a better way with a game engine (better use the engine threadpool/frame allocator instead of spawning X more threads/allocate more memory).

The biggest struggle I had with Jolt was to understand what I could and couldn't do with the bodies, sure I could create them and get a JPH::Body pointer, but then I have to use the body interface which lacks some features (like for example setting a new mass for the body) and duplicate some, I though keeping a JPH::Body was unsafe and that all accesses needed to be using the body interface, it took me some days and source code analysis to really understand that the body interface and lock system comes from the fact that Jolt is designed to be used in a multithreaded code, where multiple threads can affect bodies outside of a physics step.

Some other thing that weren't clear was raycasting, with Newton/Bullet casting a ray/shape is pretty straightforward, but with Jolt you have multiple ways of doing it, using the broadphase or the narrow phase, which should I use? Also why isn't there the hit normal like with other engines?

Of course now I understand (broadphase queries are faster but gives less info and the surface normal can be queried from the shape with the subshape ID, so that it's computed only when required).

Now on how this could be improved (aka my 2cents), maybe have small pages wiki dedicated to common stuff "how to cast a ray?", "how to setup a body and replicate it to the graphics part?", maybe add a "Get started" part to the README giving the docs links and maybe a link to GH discussions (I'm saying that because I'm not used to it).

That being said, Jolt is really amazing, thanks for the hard work!

I have a few code suggestions that I'll add in another discussion.

**Comment by jrouwe** [comment]
Hello,

Thanks for the feedback, that's very useful!

> After a few days of experimenting with Jolt it became clear this was to integrate in a better way with a game engine (better use the engine threadpool/frame allocator instead of spawning X more threads/allocate more memory).

Yes this is exactly the reason why it is a bit more complex. Initializing the allocator and threadpool are both 1 line of code (and shown in HelloWorld.cpp) so I was hoping this is not too difficult.

For your other points: I'll update the documentation as you suggest.

Thanks again!

**Comment by jrouwe** [comment]
I added some documentation on the use of body pointers and callbacks in 24147a6936c63f4f09c6c37173169d8f366f9fb4.

I noticed that I already had documentation about the use of the narrow vs the broad phase (but maybe it's indeed missing some simple ray cast example).
